staging/src/k8s.io/apimachinery/pkg/runtime/interfaces.go

```diff
@@ -155,6 +155,15 @@ type NegotiatedSerializer interface {
 	DecoderToVersion(serializer Decoder, gv GroupVersioner) Decoder
 }
 
+// ClientNegotiator handles turning an HTTP content type into the appropriate encoder.
+// Use NewClientNegotiator or NewVersionedClientNegotiator to create this interface from
+// a NegiotatedSerializer.
+type ClientNegotiator interface {
+	Encoder(contentType string, params map[string]string) (Encoder, error)
```

Comment:
Should be simple (params is the charset).  That's technically the naming that the RFC uses, will clarify.

It should return a negotiation (content type not acceptable) error.  Will clarify.

Reply:
That's handled in `NegotiateOutputMediaType`.

Reply:
Added an Error type

Reply:
> optional mediaType parameters

does that mean all parameters must be considered optional? how does that reconcile with our use of them to describe alternate GVKs?

Reply:
would be good to describe supported params or behavior if unknown params are passed

Reply:
Servers don't return media type parameters today (will double check).  It's likely if they did other clients would break.  I think the best we can do is to say "it is up to the client negotiator whether to treat these as optional or required, but current implementations consider them optional and modifiers to the base content type.

Reply:
not super satisfying, but after digging through I agree it's not worse than today, and it's not clear we could safely start returning parameters from the server